src/hdmf/data_utils.py

```diff
@@ -3,10 +3,45 @@
 import numpy as np
 from warnings import warn
 import copy
+import h5py
 
 from .utils import docval, getargs, popargs, docval_macro, get_data_shape
 
 
+def append_data(data, arg):
+    if isinstance(data, (list, DataIO)):
+        data.append(arg)
+        return data
+    elif isinstance(data, np.ndarray):
+        return np.append(data, [arg])
+    elif isinstance(data, h5py.Dataset):
```

Comment:
When would this happen? DataIO is used on write. The only time you use an `h5py.Dataset` on write is if you are making a copy to another file, in which case, you shouldn't be altering it.

Reply:
The motivation for this is to be able to add rows to dynamic tables post-write. As the test is showing on line 1314, this would involve reading a dataset in in append mode and altering the datasets by adding rows to the data, the table ids, and any index columns

Reply:
But at that point isn’t add_row just operating on the VectorData object?

Reply:
Yes. VectorData inherits from Data and Data calls this function. This is invoked when appending to a column of a file that was read in append mode. See https://github.com/hdmf-dev/hdmf/blob/2544eb9528a745a89210d3644aa77b171a8779ec/tests/unit/common/test_table.py#L1314-L1325